homework/thinhbka/lesson5-graph/1129. Shortest Path with Alternating Colors.py

```diff
@@ -0,0 +1,47 @@
+
+oo = int(1e8)
+
+class Solution:
+    def shortestAlternatingPaths(self, n: int, redEdges, blueEdges):
+        distance = [[oo,oo] for _ in range(n)]
+        visited = [[False,False] for _ in range(n)]
+        graph = [[] for _ in range(n)]# [[(1,red),(1,blue)]]
+        RED = 0
+        for edge in redEdges:
+            u = edge[0]
+            v = edge[1]
+            graph[u].append((v,RED))
+        for edge in blueEdges:
+            u = edge[0]
+            v = edge[1]
+            graph[u].append((v,1-RED))
+        front = ((0,0),(0,1))
+        distance[0][RED] = 0
+        distance[0][1-RED] = 0 
+        while front:
+            next = []
+            for node,color in front: 
+                visited[node][color] = True
+                for neighbor in graph[node]:
+                    if neighbor[1]+color == 1 and visited[neighbor[0]][neighbor[1]] is False:
+                        if distance[neighbor[0]][neighbor[1]] >=distance[node][color]+1:
```

Comment:
i dont think we need this condition. If it is not visited, it is +oo

Reply:
In this code. We need that condition. **I only mark a node is visited after i take it out of the queue** and you can see the same with the (0,RED) and (0,1-RED) at the first step when i initialized queue. So if i changed that condition where you mentioned about (```if distance[neighbor[0]][neighbor[1]] >=distance[node][color]+1:```) into ```if True:```. It failed in case:   
```
5,
[[2,2],[0,1],[0,3],[0,0],[0,4],[2,1],[2,0],[1,4],[3,4]],
[[1,3],[0,0],[0,3],[4,2],[1,0]]
```
Because node 0 has 2 red and blue edges point to itself. 
```py
        distance[0][RED] = 0
        distance[0][1-RED] = 0 
        while front:
            next = []
            for node,color in front: 
                visited[node][color] = True
                for neighbor in graph[node]:
                    if neighbor[1]+color == 1 and visited[neighbor[0]][neighbor[1]] is False:
                        if distance[neighbor[0]][neighbor[1]] >= distance[node][color]+1:
                            distance[neighbor[0]][neighbor[1]] = distance[node][color]+1
                        next.append(neighbor)
            front = next
```

Or maybe we dont need that condition if we mark a node as visited **before** we push into the queue
```py

oo = int(1e8)

oo = int(1e8)
class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges, blueEdges):
        distance[0][RED] = 0
        distance[0][1-RED] = 0 
        visited[0][RED]=True
        visited[0][1-RED]=True
        while front:
            next = []
            for node,color in front: 
                for neighbor in graph[node]:#neighbor[0]->next_node, neighbor[1]-> color of this next node
                    if neighbor[1]+color == 1 and visited[neighbor[0]][neighbor[1]] is False:
                            distance[neighbor[0]][neighbor[1]] = distance[node][color]+1
                        next.append(neighbor)
                        visited[neighbor[0]][neighbor[1]] = True


(still AC)
```
So i think the problem is the order of marking a node as visited node **before or after** deque it from queue

Reply:
This is so interesting 👯